boot/freeldr/freeldr/arch/i386/multiboot.S

```diff
@@ -98,12 +111,34 @@ mb3:
     /* Relocate itself to lower address */
     mov esi, INITIAL_BASE
     mov edi, FREELDR_BASE
+#ifndef _USE_ML
     mov ecx, offset __bss_start__ - FREELDR_BASE
+#else
+    /* Address the image with es segment */
+    // mov ax, FREELDR_PE_BASE / 16 /* == &__ImageBase */
+    // mov es, ax
+    mov eax, offset ___ImageBase
+    shr eax, 4
+    mov es, eax
+
+    /* Get address of optional header */
+    mov eax, dword ptr es:[IMAGE_DOS_HEADER_e_lfanew]
```

Comment:
Why not much simpler:
```
lea eax, dword ptr cs:[___ImageBase]
add eax, dword ptr cs:[eax + IMAGE_DOS_HEADER_e_lfanew]
mov ecx, dword ptr cs:[eax + 4 + IMAGE_FILE_HEADER_SIZE + IMAGE_OPTIONAL_HEADER_SizeOfImage]
```

Reply:
I copy-pasted and adapted slightly the code from https://git.reactos.org/?p=reactos.git;a=blob;f=boot/freeldr/freeldr/arch/realmode/i386.S;hb=d27286655018c15688afe2dd88f1ac79dbc72aa9#l36
(the same snippet also exists in amd64.S).
Your solution is more elegant, and may be ported back to the above-mentioned files too.

Reply:
Real mode code cannot address more than 64k, so you need the segment addresing there.